### PR TITLE
MQTT parser - v2

### PR DESCRIPTION
**Pull request by satta**
Previous PR: #4733 

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:  [#3549](https://redmine.openinfosecfoundation.org/issues/3549)

Describe changes (to previous PR):
- Move changes to Rust JSON code into separate commit.
- Use macros (`FAIL`, `PASS`, ...) in unit tests.
- Use `DetectU32Parse()` in protocol version rules.
- Use Rust constant `MQTT_DEFAULT_PROTOCOL_VERSION` to refer to the default version 3.
- Use pointers to scalar types as `SigMatchCtx` if only single values need to be passed. This eliminates the practice of wrapping these values into single-item structs as before.
- Be more stringent in the probing parser: try parsing the incoming data as a message and return `ALPROTO_FAILED` if that fails.

Please see previous PR for comments about things like suricata-verify tests etc.

### CODE REVIEW

File: rust/src/mqtt/mqtt.rs [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+            tmp.as_slice()
+        };
+
+        while current.len() > 0 {
+            SCLogDebug!("response: handling {}", current.len());
+            match parse_message(current, self.protocol_version) {
+                Ok((rem, msg)) => {
+                    SCLogDebug!("replacing current ({}) with rem ({})", current.len(), rem.len());
+                    current = rem;
+                    let tx = self.new_tx(msg);
```
Comment (victorjulien) [inline rust/src/mqtt/mqtt.rs:188]: it looks like there is no attempt to match requests and responses in a single transaction?

Comment (satta) [reply rust/src/mqtt/mqtt.rs:188]: Correct, the current implementation treats each single message as its own transaction. Linking request/responses when using higher QoS levels is something was thinking of (see my previous PR) but I'm not sure whether this adds a lot of value atm.

Comment (victorjulien) [reply rust/src/mqtt/mqtt.rs:188]: In general, Suricata expects transactions to be bidirectional (to be a transaction instead of merely a record). There are some parsers that don't do this, and they'll have issues pretty soon with some upcoming changes around conditional logging. So a bidirectional transaction should normally be the default.

Comment (satta) [reply rust/src/mqtt/mqtt.rs:188]: Okay. As I said, this will only be relevant for request QoS levels that are higher than 0 ("fire and forget") which do not have a response. But I will look into adding this. My approach would be to have multiple sub-objects in the log item, one for each message in the transaction, e.g.:
```json
"mqtt": {
  "retain": false,
  "dup": false,
  "publish": {
    "qos": 1,
    "topic": "topic",
    "message_id": 1,
    "message": "baa baa sheep"
  },
  "puback": {
    "qos": 0,
    "message_id": 1,
    "reason_code": 16
  }
}
```

File: rust/src/mqtt/mqtt_message.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            crate::mqtt::mqtt_message::MQTTOperation::PUBCOMP(_) => "PUBCOMP",
+            crate::mqtt::mqtt_message::MQTTOperation::SUBSCRIBE(_) => "SUBSCRIBE",
+            crate::mqtt::mqtt_message::MQTTOperation::SUBACK(_) => "SUBACK",
+            crate::mqtt::mqtt_message::MQTTOperation::UNSUBSCRIBE(_) => "UNSUBSCRIBE",
+            crate::mqtt::mqtt_message::MQTTOperation::UNSUBACK(_) => "UNSUBACK",
+            crate::mqtt::mqtt_message::MQTTOperation::PINGREQ => "PINGREQ",
+            crate::mqtt::mqtt_message::MQTTOperation::PINGRESP => "PINGRESP",
+            crate::mqtt::mqtt_message::MQTTOperation::DISCONNECT(_) => "DISCONNECT",
+            crate::mqtt::mqtt_message::MQTTOperation::AUTH(_) => "AUTH",
+            _ => "UNKNOWN",
```
Comment (catenacyber) [inline rust/src/mqtt/mqtt_message.rs:155]: Could we add the self.op number like UNKNOWN-%d ?

File: src/detect-mqtt-connack-sessionpresent.c [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+/**
+ * \internal
+ * \brief this function will free memory associated with DetectMQTTConnackSessionPresentData
+ *
+ * \param de pointer to DetectMQTTConnackSessionPresentData
+ */
+void DetectMQTTConnackSessionPresentFree(DetectEngineCtx *de_ctx, void *de_ptr)
+{
+    bool *de = (bool *)de_ptr;
+    if(de) SCFree(de);
```
Comment (catenacyber) [inline src/detect-mqtt-connack-sessionpresent.c:215]: I do not think you need to test for NULL as `SCFree` does it already

File: src/detect-mqtt-connack-sessionpresent.c [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+/**
+ * \internal
+ * \brief this function will free memory associated with DetectMQTTConnackSessionPresentData
+ *
+ * \param de pointer to DetectMQTTConnackSessionPresentData
+ */
+void DetectMQTTConnackSessionPresentFree(DetectEngineCtx *de_ctx, void *de_ptr)
+{
+    bool *de = (bool *)de_ptr;
```
Comment (catenacyber) [inline src/detect-mqtt-connack-sessionpresent.c:214]: Cast is not needed

File: src/detect-mqtt-flags.c [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        SCLogError(SC_ERR_PCRE_MATCH, "invalid flag definition: %s", rawstr);
+        return NULL;
+    }
+
+    de = SCMalloc(sizeof(DetectMQTTFlagsData));
+    if (unlikely(de == NULL))
+        return NULL;
+    de->retain = de->dup = DONT_CARE;
+
+    char copy[strlen(rawstr)+1];
```
Comment (catenacyber) [inline src/detect-mqtt-flags.c:174]: Following code looks complex and without comments

File: src/detect-mqtt-flags.c [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+ */
+static int MQTTFlagsTestParse04 (void)
+{
+    DetectMQTTFlagsData *de = NULL;
+    de = DetectMQTTFlagsParse("dup,!");
+    if (de) {
+        DetectMQTTFlagsFree(NULL, de);
+        FAIL;
+    }
+
```
Comment (catenacyber) [inline src/detect-mqtt-flags.c:347]: What about a test with the signature "dup,!dup" ?

File: src/detect-mqtt-qos.c [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+ */
+static uint8_t *DetectMQTTQosParse(const char *rawstr)
+{
+    uint8_t *de = NULL;
+#define MAX_SUBSTRINGS 30
+    int ret = 0;
+    uint8_t val;
+    int ov[MAX_SUBSTRINGS];
+
+    ret = DetectParsePcreExec(&parse_regex, rawstr, 0, 0, ov, MAX_SUBSTRINGS);
```
Comment (catenacyber) [inline src/detect-mqtt-qos.c:137]: PCRE seems an overkill here

File: suricata.yaml.in [hunk 8]
```diff
@@ -717,6 +718,8 @@ app-layer:
       enabled: yes
       detection-ports:
         dp: 5900, 5901, 5902, 5903, 5904, 5905, 5906, 5907, 5908, 5909
+    mqtt:
+      enabled: yes
```
Comment (catenacyber) [inline suricata.yaml.in:722]: Do we want to enable it by default right away ?

File: doc/userguide/output/eve/eve-json-format.rst [hunk 9]
```diff
@@ -1287,3 +1287,315 @@ Example of RFB logging, with full VNC style authentication parameters:
...
+      "flags": {
+        "username": true,
+        "password": true,
+        "will_retain": false,
+        "will": true,
+        "clean_session": true
+      },
+      "client_id": "client",
+      "username": "user",
+      "password": "pass",
```
[the remark this answers is not in the document]
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:1351]: I think that is doable. Thanks for the idea.

File: doc/userguide/output/eve/eve-json-format.rst [hunk 10]
```diff
@@ -1287,3 +1287,315 @@ Example of RFB logging, with full VNC style authentication parameters:
...
+
+::
+
+  "mqtt": {
+    "message_type": "AUTH",
+    "qos": 0,
+    "retain": false,
+    "dup": false,
+    "auth": {
+      "reason_code": 16
```
[the remark this answers is not in the document]
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:1576]: Yes, but they vary between message types. Do you think it would be a good idea to also have a human-readable version in the output? I wasn't so sure because that would inflate the output quite a bit, and for just for doing searches on log output the integer codes would work as well (see examples in documentation).

File: rust/src/mqtt/mqtt_message.rs [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+#[derive(Debug)]
+pub struct MQTTMessage {
+    pub header: FixedHeader,
+    pub op: MQTTOperation,
+}
+
+#[derive(Debug)]
+pub enum MQTTOperation {
+    UNKNOWN,
```
[the remark this answers is not in the document]
Comment (satta) [reply rust/src/mqtt/mqtt_message.rs:31]: This is just a catchall item for cases not described in the specification. Actually this only relates to (unassigned) message type 0 -- all others are handled in `parse_message()`.

File: rust/src/mqtt/mqtt_message.rs (hunk 2 above)
Comment (satta) [reply rust/src/mqtt/mqtt_message.rs:155]: This is always 0, so I don't really see a point.

File: src/detect-mqtt-connack-sessionpresent.c (hunk 4 above)
Comment (satta) [reply src/detect-mqtt-connack-sessionpresent.c:214]: Okay, thanks for the hint.

File: src/detect-mqtt-connack-sessionpresent.c (hunk 3 above)
Comment (satta) [reply src/detect-mqtt-connack-sessionpresent.c:215]: Ah, this is probably just a habit :) Will keep this in mind.

File: rust/src/mqtt/detect.rs [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_has_topic(
+    tx: &MQTTTransaction,
+    result: *mut bool,
+    topic: *const std::os::raw::c_char,
+) -> u8 {
+    *result = false;
+    match tx.msg.op {
+        MQTTOperation::SUBSCRIBE(ref subv) => {
+            let c_topic = CStr::from_ptr(topic).to_str().unwrap();
```
[the remark this answers is not in the document]
Comment (satta) [reply rust/src/mqtt/detect.rs:256]: Probably not. Good catch. I will grep the whole MQTT code for possible use of `unwrap()` and check whether any of these needs extra attention.

File: src/detect-mqtt-flags.c (hunk 6 above)
Comment (satta) [reply src/detect-mqtt-flags.c:347]: Good idea. There's also code missing to detect this contradictory situation. Will add code to handle this case.

File: src/detect-mqtt-qos.c (hunk 7 above)
Comment (satta) [reply src/detect-mqtt-qos.c:137]: True. Removed that here and just worked with the parsed integer value.

File: suricata.yaml.in (hunk 8 above)
Comment (satta) [reply suricata.yaml.in:722]: I'm fine with both. What do the others think?

File: doc/userguide/output/eve/eve-json-format.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply doc/userguide/output/eve/eve-json-format.rst:1576]: Ok, maybe leave a comment about this

File: src/detect-mqtt-connack-sessionpresent.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-mqtt-connack-sessionpresent.c:215]: That is a good habit ;-)

File: rust/src/mqtt/mqtt_message.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/mqtt/mqtt_message.rs:155]: Ok, maybe add this comment
Rust `_` does not show that this can only be 0 :-)

File: rust/src/mqtt/parser.rs [hunk 13]
```diff
@@ -0,0 +1,628 @@
...
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                _ => {
+                    let msg = MQTTMessage {
+                        header: header,
+                        op: MQTTOperation::UNKNOWN, // should not happen
+                    };
```
Comment (catenacyber) [inline rust/src/mqtt/parser.rs:616]: can you panic when in fuzz mode if this is not supposed to happen ?

Comment (satta) [reply rust/src/mqtt/parser.rs:616]: Let me rephrase: It is still possible for this case to be triggered with broken traffic (message type set to reserved value 0). I wouldn't panic here since this does not indicate a programming error but rather an anomaly in traffic. One might, for example, raise an event ("use of reserved message type 0") here. Ok, I can add that.

File: doc/userguide/output/eve/eve-json-format.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:1576]: In the RST documentation? I have already mentioned that one should refer to the specification for the meaning of the reason code values (https://github.com/OISF/suricata/pull/4934/files#diff-7823d9d82df45a9858e8925a6259022eR1438, https://github.com/OISF/suricata/pull/4934/files#diff-7823d9d82df45a9858e8925a6259022eR1563 etc.).

File: rust/src/mqtt/parser.rs (hunk 13 above)
Comment (catenacyber) [reply rust/src/mqtt/parser.rs:616]: Ok for the event if it is not covered by another signature like `mqtt.type:0` .
I would modify the comment to say that this should not happen with normal traffic.
The other comments `should not happen` you made seem unreachable code

File: doc/userguide/output/eve/eve-json-format.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply doc/userguide/output/eve/eve-json-format.rst:1576]: Thanks, that seems fair enough :-)

File: rust/src/mqtt/parser.rs (hunk 13 above)
Comment (satta) [reply rust/src/mqtt/parser.rs:616]: Okay, right.

Comment (victorjulien) [reply rust/src/mqtt/parser.rs:616]: FYI, we're adding `debug_validate_bug_on` which will operate the same as `DEBUG_VALIDATE_BUG_ON` in our C code.

But it sounds like it's not appropriate here.

File: rust/src/mqtt/mqtt.rs (hunk 1 above)
Comment (satta) [reply rust/src/mqtt/mqtt.rs:188]: This might take a while as it will also impact almost all MQTT detection code, just FYI.

File: doc/userguide/output/eve/eve-json-format.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:1351]: Done. Will be in next PR.

[the remark this answers is not in the document]
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:1351]: FYI: This will only be configurable in the first iteration for `event_type: mqtt` log output. There is one small exception: I have also implemented MQTT metadata output in `alert` events, but since I could not find a way to access the configuration items of another type in the same EVE output, I had to settle for a default, which is to log passwords. 
I took some inspiration from the DNS code, which also falls back to a default setting of "output all rrtypes" in alerts, regardless of what is configured in the `dns` type.

File: suricata.yaml.in (hunk 8 above)
Comment (satta) [reply suricata.yaml.in:722]: Disabled it for now by default, since it is so easy to turn back on.

File: rust/src/mqtt/mqtt.rs (hunk 1 above)
Comment (satta) [reply rust/src/mqtt/mqtt.rs:188]: Phew, done. In next MR.

**Comment by satta** [comment]
New PR: #4960
